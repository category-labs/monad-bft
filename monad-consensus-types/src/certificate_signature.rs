use monad_crypto::{
    secp256k1::{Error as SecpError, KeyPair as SecpKeyPair, PubKey as SecpPubKey, SecpSignature},
    NopSignature,
};

use crate::{message_signature::MessageSignature, validation::Hashable};

pub trait CertificateKeyPair: Send + Sized + Sync + 'static {
    type PubKeyType: std::cmp::Eq + std::hash::Hash + Copy;
    type Error: std::error::Error + Send + Sync;

    fn from_bytes(secret: impl AsMut<[u8]>) -> Result<Self, Self::Error>;
    fn pubkey(&self) -> Self::PubKeyType;
}

pub trait CertificateSignature:
    Copy + Clone + Eq + Hashable + Send + Sync + std::fmt::Debug + 'static
{
    type KeyPairType: CertificateKeyPair;
    type Error: std::error::Error + Send + Sync;

    fn sign(msg: &[u8], keypair: &Self::KeyPairType) -> Self;
    fn verify(
        &self,
        msg: &[u8],
        pubkey: &<Self::KeyPairType as CertificateKeyPair>::PubKeyType,
    ) -> Result<(), Self::Error>;

    fn serialize(&self) -> Vec<u8>;
    fn deserialize(signature: &[u8]) -> Result<Self, Self::Error>;
}

pub trait CertificateSignatureRecoverable: CertificateSignature {
    fn recover_pubkey(
        &self,
        msg: &[u8],
    ) -> Result<
        <<Self as CertificateSignature>::KeyPairType as CertificateKeyPair>::PubKeyType,
        <Self as CertificateSignature>::Error,
    >;
}

impl CertificateKeyPair for SecpKeyPair {
    type PubKeyType = SecpPubKey;
    type Error = SecpError;

    fn from_bytes(secret: impl AsMut<[u8]>) -> Result<Self, Self::Error> {
        Self::from_bytes(secret)
    }

    fn pubkey(&self) -> Self::PubKeyType {
        self.pubkey()
    }
}

impl Hashable for SecpSignature {
    fn hash<H: crate::validation::Hasher>(&self, state: &mut H) {
        let slice = unsafe { std::mem::transmute::<Self, [u8; 65]>(*self) };
        state.update(slice)
    }
}

impl CertificateSignature for SecpSignature {
    type KeyPairType = SecpKeyPair;
    type Error = SecpError;

    fn sign(msg: &[u8], keypair: &Self::KeyPairType) -> Self {
        keypair.sign(msg)
    }

    fn verify(
        &self,
        msg: &[u8],
        pubkey: &<Self::KeyPairType as CertificateKeyPair>::PubKeyType,
    ) -> Result<(), Self::Error> {
        pubkey.verify(msg, self)
    }

    fn serialize(&self) -> Vec<u8> {
        self.serialize().to_vec()
    }

    fn deserialize(signature: &[u8]) -> Result<Self, Self::Error> {
        Self::deserialize(signature)
    }
}

impl CertificateSignatureRecoverable for SecpSignature {
    fn recover_pubkey(
        &self,
        msg: &[u8],
    ) -> Result<
        <<Self as CertificateSignature>::KeyPairType as CertificateKeyPair>::PubKeyType,
        <Self as CertificateSignature>::Error,
    > {
        self.recover_pubkey(msg)
    }
}

impl Hashable for NopSignature {
    fn hash<H: crate::validation::Hasher>(&self, state: &mut H) {
        let slice = unsafe { std::mem::transmute::<Self, [u8; 72]>(*self) };
        state.update(slice)
    }
}

impl CertificateSignature for NopSignature {
    type KeyPairType = SecpKeyPair;
    type Error = SecpError;

    fn sign(msg: &[u8], keypair: &Self::KeyPairType) -> Self {
        <Self as MessageSignature>::sign(msg, keypair)
    }

    fn verify(
        &self,
        msg: &[u8],
        pubkey: &<Self::KeyPairType as CertificateKeyPair>::PubKeyType,
    ) -> Result<(), Self::Error> {
        <Self as MessageSignature>::verify(self, msg, pubkey)
    }

    fn serialize(&self) -> Vec<u8> {
        <Self as MessageSignature>::serialize(self)
    }

    fn deserialize(signature: &[u8]) -> Result<Self, Self::Error> {
        <Self as MessageSignature>::deserialize(signature)
    }
}

impl CertificateSignatureRecoverable for NopSignature {
    fn recover_pubkey(
        &self,
        msg: &[u8],
    ) -> Result<
        <<Self as CertificateSignature>::KeyPairType as CertificateKeyPair>::PubKeyType,
        <Self as CertificateSignature>::Error,
    > {
        <Self as MessageSignature>::recover_pubkey(self, msg)
    }
}

#[cfg(test)]
mod test {
    use std::ops::AddAssign;

    use monad_crypto::secp256k1::SecpSignature;

    // valid certificate signature tests
    use crate::certificate_signature::{
        CertificateKeyPair, CertificateSignature, CertificateSignatureRecoverable,
    };
    macro_rules! test_all_certificate_signature {
        ($test_name:ident, $test_code:block) => {
            mod $test_name {
                use monad_crypto::{secp256k1::SecpSignature, NopSignature};

                use super::*;

                fn invoke<T>()
                where
                    T: CertificateSignature + std::fmt::Debug,
                    <T::KeyPairType as CertificateKeyPair>::PubKeyType:
                        std::cmp::Eq + std::fmt::Debug,
                {
                    $test_code
                }

                #[test]
                fn secpsignature() {
                    invoke::<SecpSignature>();
                }

                #[test]
                fn nopsignature() {
                    invoke::<NopSignature>();
                }

                // TODO: add module for bls signature
            }
        };
    }

    macro_rules! test_all_certificate_signature_recoverable {
        ($test_name:ident, $test_code:block) => {
            mod $test_name {
                use monad_crypto::{secp256k1::SecpSignature, NopSignature};

                use super::*;

                fn invoke<T>()
                where
                    T: CertificateSignatureRecoverable + std::fmt::Debug,
                    <T::KeyPairType as CertificateKeyPair>::PubKeyType:
                        std::cmp::Eq + std::fmt::Debug,
                {
                    $test_code
                }

                #[test]
                fn secpsignature() {
                    invoke::<SecpSignature>();
                }

                #[test]
                fn nopsignature() {
                    invoke::<NopSignature>();
                }
            }
        };
    }

    test_all_certificate_signature!(test_keypair_deterministic_creation, {
        let mut s1 = [127_u8; 32];
        let mut s2 = [127_u8; 32];

        assert_eq!(s1, s2);

        let k1 = <T::KeyPairType as CertificateKeyPair>::from_bytes(s1.as_mut_slice()).unwrap();
        let k2 = <T::KeyPairType as CertificateKeyPair>::from_bytes(s2.as_mut_slice()).unwrap();

        assert_eq!(k1.pubkey(), k2.pubkey());
    });

    test_all_certificate_signature!(test_serialization_roundtrip, {
        let mut s = [127_u8; 32];
        let certkey = <T::KeyPairType as CertificateKeyPair>::from_bytes(s.as_mut_slice()).unwrap();

        let msg = b"hello world";
        let sig = T::sign(msg, &certkey);

        let sig_bytes = sig.serialize();
        let sig_de = T::deserialize(sig_bytes.as_ref()).unwrap();

        assert_eq!(sig, sig_de);
    });

    test_all_certificate_signature!(test_signature_verify, {
        let mut s = [127_u8; 32];
        let certkey = <T::KeyPairType as CertificateKeyPair>::from_bytes(s.as_mut_slice()).unwrap();

        let msg = b"hello world";
        let sig = T::sign(msg, &certkey);

        assert!(sig.verify(msg, &certkey.pubkey()).is_ok());
    });

    test_all_certificate_signature_recoverable!(test_recover, {
        let mut s = [127_u8; 32];
        let certkey = <T::KeyPairType as CertificateKeyPair>::from_bytes(s.as_mut_slice()).unwrap();

        let msg = b"hello world";
        let sig = T::sign(msg, &certkey);

        assert_eq!(sig.recover_pubkey(msg).unwrap(), certkey.pubkey());
    });

    // invalid certificate signature tests
    #[test]
    fn test_verify_error() {
        let mut s = [127_u8; 32];
        let certkey = <<SecpSignature as CertificateSignature>::KeyPairType as CertificateKeyPair>::from_bytes(s.as_mut_slice()).unwrap();

        let msg = b"hello world";
        let invalid_msg = b"bye world";
        let sig = <SecpSignature as CertificateSignature>::sign(msg, &certkey);

        assert!(<SecpSignature as CertificateSignature>::verify(
            &sig,
            invalid_msg,
            &certkey.pubkey()
        )
        .is_err());
    }

    #[test]
    fn test_deser_error() {
        let mut s = [127_u8; 32];
        let certkey = <<SecpSignature as CertificateSignature>::KeyPairType as CertificateKeyPair>::from_bytes(s.as_mut_slice()).unwrap();

        let msg = b"hello world";
        let sig = <SecpSignature as CertificateSignature>::sign(msg, &certkey);

        let mut sig_bytes = <SecpSignature as CertificateSignature>::serialize(&sig);

        // the last byte is the recoveryId
        // recoveryId is 0..=3, adding 4 makes it invalid
        sig_bytes.last_mut().unwrap().add_assign(5);

        assert!(<SecpSignature as CertificateSignature>::deserialize(&sig_bytes).is_err());
    }
}
