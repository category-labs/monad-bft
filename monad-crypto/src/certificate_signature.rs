use std::{
    collections::hash_map::DefaultHasher,
    fmt::{Debug, Display},
    hash::{Hash, Hasher},
};

use crate::{
    bls12_381::{BlsError, BlsKeyPair, BlsPubKey, BlsSignature},
    hasher::Hashable,
    secp256k1::{Error as SecpError, KeyPair as SecpKeyPair, PubKey as SecpPubKey, SecpSignature},
    NopKeyPair, NopPubKey, NopSignature,
};

pub trait PubKey:
    Debug + Eq + Hash + Ord + PartialOrd + Copy + Send + Sync + Unpin + 'static
{
    type Error: Display + Debug + Send + Sync;
    fn from_bytes(pubkey: &[u8]) -> Result<Self, Self::Error>;
    fn bytes(&self) -> Vec<u8>;
}

pub trait CertificateKeyPair: Send + Sized + Sync + 'static {
    type PubKeyType: PubKey;
    type Error: Display + Debug + Send + Sync;

    fn from_bytes(secret: &mut [u8]) -> Result<Self, Self::Error>;
    fn pubkey(&self) -> Self::PubKeyType;
}

pub type CertificateSignaturePubKey<T> =
    <<T as CertificateSignature>::KeyPairType as CertificateKeyPair>::PubKeyType;

pub trait CertificateSignature:
    Copy + Clone + Eq + Hashable + Send + Sync + std::fmt::Debug + std::hash::Hash + 'static
{
    type KeyPairType: CertificateKeyPair;
    type Error: Display + Debug + Send + Sync;

    fn sign(msg: &[u8], keypair: &Self::KeyPairType) -> Self;
    fn verify(
        &self,
        msg: &[u8],
        pubkey: &CertificateSignaturePubKey<Self>,
    ) -> Result<(), Self::Error>;

    fn serialize(&self) -> Vec<u8>;
    fn deserialize(signature: &[u8]) -> Result<Self, Self::Error>;
}

pub trait CertificateSignatureRecoverable: CertificateSignature {
    fn recover_pubkey(
        &self,
        msg: &[u8],
    ) -> Result<CertificateSignaturePubKey<Self>, <Self as CertificateSignature>::Error>;
}

impl PubKey for SecpPubKey {
    type Error = SecpError;

    fn from_bytes(pubkey: &[u8]) -> Result<Self, Self::Error> {
        Self::from_slice(pubkey)
    }

    fn bytes(&self) -> Vec<u8> {
        SecpPubKey::bytes(self)
    }
}

impl CertificateKeyPair for SecpKeyPair {
    type PubKeyType = SecpPubKey;
    type Error = SecpError;

    fn from_bytes(secret: &mut [u8]) -> Result<Self, Self::Error> {
        Self::from_bytes(secret)
    }

    fn pubkey(&self) -> Self::PubKeyType {
        self.pubkey()
    }
}

impl CertificateSignature for SecpSignature {
    type KeyPairType = SecpKeyPair;
    type Error = SecpError;

    fn sign(msg: &[u8], keypair: &Self::KeyPairType) -> Self {
        keypair.sign(msg)
    }

    fn verify(
        &self,
        msg: &[u8],
        pubkey: &CertificateSignaturePubKey<Self>,
    ) -> Result<(), Self::Error> {
        pubkey.verify(msg, self)
    }

    fn serialize(&self) -> Vec<u8> {
        self.serialize().to_vec()
    }

    fn deserialize(signature: &[u8]) -> Result<Self, Self::Error> {
        Self::deserialize(signature)
    }
}

impl PubKey for NopPubKey {
    type Error = &'static str;

    fn from_bytes(pubkey: &[u8]) -> Result<Self, Self::Error> {
        Ok(Self(
            pubkey
                .try_into()
                .map_err(|_| "couldn't deserialize pubkey")?,
        ))
    }

    fn bytes(&self) -> Vec<u8> {
        self.0.to_vec()
    }
}

impl CertificateKeyPair for NopKeyPair {
    type PubKeyType = NopPubKey;
    type Error = &'static str;

    fn from_bytes(secret: &mut [u8]) -> Result<Self, Self::Error> {
        Ok(Self {
            pubkey: NopPubKey::from_bytes(secret)?,
        })
    }

    fn pubkey(&self) -> Self::PubKeyType {
        self.pubkey
    }
}

impl CertificateSignatureRecoverable for SecpSignature {
    fn recover_pubkey(
        &self,
        msg: &[u8],
    ) -> Result<CertificateSignaturePubKey<Self>, <Self as CertificateSignature>::Error> {
        self.recover_pubkey(msg)
    }
}

impl CertificateSignature for NopSignature {
    type KeyPairType = NopKeyPair;
    type Error = &'static str;

    fn sign(msg: &[u8], keypair: &Self::KeyPairType) -> Self {
        let mut hasher = DefaultHasher::new();
        hasher.write(msg);

        NopSignature {
            pubkey: keypair.pubkey,
            id: hasher.finish(),
        }
    }

    fn verify(
        &self,
        _msg: &[u8],
        pubkey: &CertificateSignaturePubKey<Self>,
    ) -> Result<(), Self::Error> {
        if &self.pubkey == pubkey {
            Ok(())
        } else {
            Err("invalid pubkey")
        }
    }

    fn serialize(&self) -> Vec<u8> {
        self.id
            .to_le_bytes()
            .into_iter()
            .chain(self.pubkey.bytes())
            .collect()
    }

    fn deserialize(signature: &[u8]) -> Result<Self, Self::Error> {
        let id = u64::from_le_bytes(signature[..8].try_into().unwrap());
        let pubkey = NopPubKey::from_bytes(&signature[8..])?;
        Ok(Self { pubkey, id })
    }
}

impl CertificateSignatureRecoverable for NopSignature {
    fn recover_pubkey(
        &self,
        _msg: &[u8],
    ) -> Result<CertificateSignaturePubKey<Self>, <Self as CertificateSignature>::Error> {
        Ok(self.pubkey)
    }
}

impl PubKey for BlsPubKey {
    type Error = BlsError;

    fn from_bytes(pubkey: &[u8]) -> Result<Self, Self::Error> {
        Self::deserialize(pubkey)
    }

    fn bytes(&self) -> Vec<u8> {
        self.serialize()
    }
}

impl CertificateKeyPair for BlsKeyPair {
    type PubKeyType = BlsPubKey;
    type Error = BlsError;

    fn from_bytes(secret: &mut [u8]) -> Result<Self, Self::Error> {
        Self::from_bytes(secret)
    }

    fn pubkey(&self) -> Self::PubKeyType {
        self.pubkey()
    }
}

impl CertificateSignature for BlsSignature {
    type KeyPairType = BlsKeyPair;
    type Error = BlsError;

    fn sign(msg: &[u8], keypair: &Self::KeyPairType) -> Self {
        keypair.sign(msg)
    }

    fn verify(
        &self,
        msg: &[u8],
        pubkey: &CertificateSignaturePubKey<Self>,
    ) -> Result<(), Self::Error> {
        self.verify(msg, pubkey)
    }

    fn serialize(&self) -> Vec<u8> {
        self.serialize()
    }

    fn deserialize(signature: &[u8]) -> Result<Self, Self::Error> {
        Self::deserialize(signature)
    }
}

#[cfg(test)]
mod test {
    use std::ops::AddAssign;

    // valid certificate signature tests
    use crate::certificate_signature::{
        CertificateKeyPair, CertificateSignature, CertificateSignaturePubKey,
        CertificateSignatureRecoverable,
    };
    use crate::{bls12_381::BlsSignature, secp256k1::SecpSignature};
    macro_rules! test_all_certificate_signature {
        ($test_name:ident, $test_code:block) => {
            mod $test_name {
                use super::*;
                use crate::{secp256k1::SecpSignature, NopSignature};

                fn invoke<T>()
                where
                    T: CertificateSignature + std::fmt::Debug,
                    CertificateSignaturePubKey<T>: std::cmp::Eq + std::fmt::Debug,
                {
                    $test_code
                }

                #[test]
                fn secpsignature() {
                    invoke::<SecpSignature>();
                }

                #[test]
                fn nopsignature() {
                    invoke::<NopSignature>();
                }

                #[test]
                fn blssignature() {
                    invoke::<BlsSignature>();
                }
            }
        };
    }

    macro_rules! test_all_certificate_signature_recoverable {
        ($test_name:ident, $test_code:block) => {
            mod $test_name {
                use super::*;
                use crate::{secp256k1::SecpSignature, NopSignature};

                fn invoke<T>()
                where
                    T: CertificateSignatureRecoverable + std::fmt::Debug,
                    CertificateSignaturePubKey<T>: std::cmp::Eq + std::fmt::Debug,
                {
                    $test_code
                }

                #[test]
                fn secpsignature() {
                    invoke::<SecpSignature>();
                }

                #[test]
                fn nopsignature() {
                    invoke::<NopSignature>();
                }
            }
        };
    }

    test_all_certificate_signature!(test_keypair_deterministic_creation, {
        let mut s1 = [127_u8; 32];
        let mut s2 = [127_u8; 32];

        assert_eq!(s1, s2);

        let k1 = <T::KeyPairType as CertificateKeyPair>::from_bytes(s1.as_mut_slice()).unwrap();
        let k2 = <T::KeyPairType as CertificateKeyPair>::from_bytes(s2.as_mut_slice()).unwrap();

        assert_eq!(k1.pubkey(), k2.pubkey());
    });

    test_all_certificate_signature!(test_serialization_roundtrip, {
        let mut s = [127_u8; 32];
        let certkey = <T::KeyPairType as CertificateKeyPair>::from_bytes(s.as_mut_slice()).unwrap();

        let msg = b"hello world";
        let sig = T::sign(msg, &certkey);

        let sig_bytes = sig.serialize();
        let sig_de = T::deserialize(sig_bytes.as_ref()).unwrap();

        assert_eq!(sig, sig_de);
    });

    test_all_certificate_signature!(test_signature_verify, {
        let mut s = [127_u8; 32];
        let certkey = <T::KeyPairType as CertificateKeyPair>::from_bytes(s.as_mut_slice()).unwrap();

        let msg = b"hello world";
        let sig = T::sign(msg, &certkey);

        assert!(sig.verify(msg, &certkey.pubkey()).is_ok());
    });

    test_all_certificate_signature_recoverable!(test_recover, {
        let mut s = [127_u8; 32];
        let certkey = <T::KeyPairType as CertificateKeyPair>::from_bytes(s.as_mut_slice()).unwrap();

        let msg = b"hello world";
        let sig = T::sign(msg, &certkey);

        assert_eq!(sig.recover_pubkey(msg).unwrap(), certkey.pubkey());
    });

    // invalid certificate signature tests
    #[test]
    fn test_verify_error() {
        let mut s = [127_u8; 32];
        let certkey = <<SecpSignature as CertificateSignature>::KeyPairType as CertificateKeyPair>::from_bytes(s.as_mut_slice()).unwrap();

        let msg = b"hello world";
        let invalid_msg = b"bye world";
        let sig = <SecpSignature as CertificateSignature>::sign(msg, &certkey);

        assert!(<SecpSignature as CertificateSignature>::verify(
            &sig,
            invalid_msg,
            &certkey.pubkey()
        )
        .is_err());
    }

    #[test]
    fn test_deser_error() {
        let mut s = [127_u8; 32];
        let certkey = <<SecpSignature as CertificateSignature>::KeyPairType as CertificateKeyPair>::from_bytes(s.as_mut_slice()).unwrap();

        let msg = b"hello world";
        let sig = <SecpSignature as CertificateSignature>::sign(msg, &certkey);

        let mut sig_bytes = <SecpSignature as CertificateSignature>::serialize(&sig);

        // the last byte is the recoveryId
        // recoveryId is 0..=3, adding 4 makes it invalid
        sig_bytes.last_mut().unwrap().add_assign(5);

        assert!(<SecpSignature as CertificateSignature>::deserialize(&sig_bytes).is_err());
    }
}
