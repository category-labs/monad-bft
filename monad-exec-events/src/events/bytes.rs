// Copyright (C) 2025 Category Labs, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use std::mem::MaybeUninit;

/// Safely casts the beginning chunk of the provided byte slice `bytes` into a `&T` if possible,
/// producing the remaining half of the slice alongside in a tuple.
fn split_ref_from_bytes<T>(bytes: &'_ [u8]) -> Result<(&'_ T, &'_ [u8]), String> {
    let Some((bytes, rest)) = bytes.split_at_checked(size_of::<T>()) else {
        return Err(format!(
            "Expected slice with length at least {} but slice has length {}",
            size_of::<T>(),
            bytes.len()
        ));
    };

    assert_eq!(bytes.len(), size_of::<T>());

    let bytes_ptr = bytes.as_ptr();

    if (bytes_ptr as usize) % align_of::<T>() != 0 {
        return Err(format!(
            "Expected slice with alignment {} but slice has ptr {}",
            align_of::<T>(),
            bytes_ptr as usize
        ));
    }

    Ok((unsafe { &*(bytes_ptr as *const T) }, rest))
}

/// Safely casts the provided byte slice `bytes` into a `&T`.
pub(crate) fn ref_from_bytes<T>(bytes: &'_ [u8]) -> Result<&'_ T, String> {
    let (value_ref, rest) = split_ref_from_bytes(bytes)?;

    if !rest.is_empty() {
        return Err(format!(
            "Expected slice with length {} but slice has length {}",
            size_of::<T>(),
            bytes.len()
        ));
    }

    Ok(value_ref)
}

/// Safely casts the provided byte slice `bytes` into a `&T` and produces it along with an array of
/// trailing byte slices from sizes specified using the produced `&T`.
///
/// Some execution events specify trailing byte data, like
/// [`ExecEventsRef::TxnStart`](crate::ExecEventsRef), whose size can only be determined after first
/// parsing the event, [`monad_exec_txn_start`](crate::ffi::monad_exec_txn_start). This function
/// takes a fn which, given this `&T` reference, produces an array of trailing byte sizes which are
/// subsequently split out and produced alongisde the `&T`.
pub(crate) fn ref_from_bytes_with_trailing<'ring, T, const N: usize>(
    bytes: &'ring [u8],
    trailing_lengths: impl FnOnce(&T) -> [usize; N],
) -> Result<(&'ring T, [&'ring [u8]; N]), String> {
    let (value, mut bytes) = split_ref_from_bytes::<T>(bytes)?;

    let trailing_lengths = trailing_lengths(value);

    let mut trailing: [MaybeUninit<&'ring [u8]>; N] =
        unsafe { MaybeUninit::uninit().assume_init() };

    for (idx, length) in trailing_lengths.into_iter().enumerate() {
        let Some((bytes_next, bytes_rest)) = bytes.split_at_checked(length) else {
            return Err(format!(
                "Expected slice with length greater than {length} but slice has length {}",
                bytes.len()
            ));
        };

        trailing[idx] = MaybeUninit::new(bytes_next);
        bytes = bytes_rest;
    }

    if !bytes.is_empty() {
        return Err(format!(
            "Payload trailing slice has {} additional bytes",
            bytes.len()
        ));
    }

    let trailing = std::array::from_fn(|i| unsafe { trailing[i].assume_init() });

    Ok((value, trailing))
}
