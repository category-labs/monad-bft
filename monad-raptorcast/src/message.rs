use alloy_rlp::{encode_list, Decodable, Encodable, Header, RlpDecodable, RlpEncodable};
use bytes::{Bytes, BytesMut};
use monad_compress::{zstd::ZstdCompression, CompressionAlgo};
use monad_crypto::certificate_signature::CertificateSignatureRecoverable;
use monad_peer_discovery::PeerDiscoveryMessage;

const SERIALIZE_VERSION: u32 = 1;
// compression versions
const UNCOMPRESSED_VERSION: u32 = 1;
const DEFAULT_ZSTD_VERSION: u32 = 2;

#[derive(RlpEncodable, RlpDecodable)]
struct NetworkMessageVersion {
    pub serialize_version: u32,
    pub compression_version: u32,
}

impl NetworkMessageVersion {
    pub fn version() -> Self {
        Self {
            serialize_version: SERIALIZE_VERSION,
            compression_version: UNCOMPRESSED_VERSION,
        }
    }
}

pub enum OutboundRouterMessage<OM, ST: CertificateSignatureRecoverable> {
    AppMessage(OM),
    PeerDiscoveryMessage(PeerDiscoveryMessage<ST>),
}

#[derive(Debug)]
pub struct SerializeError(pub String);

impl<OM: Encodable, ST: CertificateSignatureRecoverable> OutboundRouterMessage<OM, ST> {
    pub fn try_serialize(self) -> Result<Bytes, SerializeError> {
        let mut buf = BytesMut::new();

        let version = NetworkMessageVersion::version();
        match self {
            Self::AppMessage(app_message) => {
                if version.compression_version == UNCOMPRESSED_VERSION {
                    // encode as uncompressed message
                    let enc: [&dyn Encodable; 3] = [&version, &1u8, &app_message];
                    encode_list::<_, dyn Encodable>(&enc, &mut buf);
                } else if version.compression_version == DEFAULT_ZSTD_VERSION {
                    // compress message
                    let mut rlp_encoded_msg = BytesMut::new();
                    app_message.encode(&mut rlp_encoded_msg);

                    let mut compressed_app_message = Vec::new();
                    ZstdCompression::default()
                        .compress(&rlp_encoded_msg, &mut compressed_app_message)
                        .map_err(|err| SerializeError(format!("compression error: {:?}", err)))?;
                    let compressed_app_message = Bytes::from(compressed_app_message);

                    // encode as compressed message
                    let enc: [&dyn Encodable; 3] = [&version, &1u8, &compressed_app_message];
                    encode_list::<_, dyn Encodable>(&enc, &mut buf);
                } else {
                    unreachable!()
                };
            }
            Self::PeerDiscoveryMessage(peer_disc_message) => {
                // encode as uncompressed message
                let enc: [&dyn Encodable; 3] = [&version, &2u8, &peer_disc_message];
                encode_list::<_, dyn Encodable>(&enc, &mut buf);
            }
        };

        Ok(buf.into())
    }
}

pub enum InboundRouterMessage<M, ST: CertificateSignatureRecoverable> {
    AppMessage(M),
    PeerDiscoveryMessage(PeerDiscoveryMessage<ST>),
}

#[derive(Debug)]
pub struct DeserializeError(pub String);

impl From<alloy_rlp::Error> for DeserializeError {
    fn from(err: alloy_rlp::Error) -> Self {
        DeserializeError(format!("rlp decode error: {:?}", err))
    }
}

impl<M: Decodable, ST: CertificateSignatureRecoverable> InboundRouterMessage<M, ST> {
    pub fn try_deserialize(data: &Bytes) -> Result<Self, DeserializeError> {
        let mut payload =
            Header::decode_bytes(&mut data.as_ref(), true).map_err(DeserializeError::from)?;

        let version =
            NetworkMessageVersion::decode(&mut payload).map_err(DeserializeError::from)?;
        let message_type = u8::decode(&mut payload).map_err(DeserializeError::from)?;
        match message_type {
            1 => {
                match version.compression_version {
                    UNCOMPRESSED_VERSION => {
                        // decode as uncompressed message
                        let app_message =
                            M::decode(&mut payload).map_err(DeserializeError::from)?;
                        Ok(Self::AppMessage(app_message))
                    }
                    DEFAULT_ZSTD_VERSION => {
                        let compressed_app_message =
                            Bytes::decode(&mut payload).map_err(DeserializeError::from)?;

                        // decompress message
                        let mut decompressed_app_message = Vec::new();
                        ZstdCompression::default()
                            .decompress(&compressed_app_message, &mut decompressed_app_message)
                            .map_err(|err| {
                                DeserializeError(format!("decompression error: {:?}", err))
                            })?;

                        let app_message = M::decode(&mut decompressed_app_message.as_ref())
                            .map_err(DeserializeError::from)?;
                        Ok(Self::AppMessage(app_message))
                    }
                    _ => Err(DeserializeError("unknown compression version".into())),
                }
            }
            2 => {
                let peer_disc_message =
                    PeerDiscoveryMessage::decode(&mut payload).map_err(DeserializeError::from)?;
                Ok(Self::PeerDiscoveryMessage(peer_disc_message))
            }
            _ => Err(DeserializeError("unknown message type".into())),
        }
    }
}
