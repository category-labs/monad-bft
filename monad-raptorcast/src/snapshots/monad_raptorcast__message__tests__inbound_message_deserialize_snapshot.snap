---
source: monad-raptorcast/src/message.rs
expression: hex_encoded_roundtrip
---
c8c2010101c3823039
